src/EditorFeatures/Core/Shared/Preview/PreviewWorkspace.cs

```diff
@@ -31,6 +32,30 @@ public PreviewWorkspace(Solution solution)
         this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.SolutionChanged, oldSolution, newSolution);
     }
 
+    public static ReferenceCountedDisposable<PreviewWorkspace> CreateWithDocumentContents(
+        TextDocument document, SourceTextContainer textContainer)
+    {
+        // Ensure the solution's view of this file is consistent across all linked files within it.
+
+        // Performance: Replace related documents to in thisworkspaces with the 
+        // relating SourceText. This prevents cascading forks as taggers call to
+        // GetOpenTextDocumentInCurrentContextWithChanges would eventually wind up
+        // calling Solution.WithDocumentText using the related ids.
+        var newSolution = document.Project.Solution.WithDocumentText(
+            document.Project.Solution.GetRelatedDocumentIds(document.Id),
+            textContainer.CurrentText,
+            PreservationMode.PreserveIdentity);
+
+        using var previewWorkspace = new ReferenceCountedDisposable<PreviewWorkspace>(new PreviewWorkspace(newSolution));
+
+        // TODO: Determine if this is necesarry.  Existing code comments mention that this is needed so that things
+        // like the LightBulb work.  But it is unclear if that's actually the case.  It is possible some features
+        // may do things slightly differently if a doc is open or not.  But those cases should be rare.
+        previewWorkspace.Target.OpenDocument(document.Id, textContainer);
```

Comment:
keeping this @jasonmalinowski as all code paths currently have it.  BUt now it is at least contained to one location someone could experiment with.